test/integration/interaction/Discovery.hpp

```diff
@@ -0,0 +1,77 @@
+#ifndef IN_TEST_DISCOVERY_HPP
+#define IN_TEST_DISCOVERY_HPP
+
+#include "BigHelloWorld.h"
+#include "Gateway.hpp"
+#include "Client.hpp"
+
+#include <uxr/client/client.h>
+#include <ucdr/microcdr.h>
+
+#include <gtest/gtest.h>
+#include <iostream>
+#include <thread>
+
+class Discovery
+{
+public:
+    Discovery(int transport, const std::vector<uint16_t>& agent_ports)
+    : agent_ports_(agent_ports)
+    , transport_(transport)
+    {
+    }
+
+    void unicast(const std::vector<uint16_t>& discovery_ports)
+    {
+        std::vector<uxrAgentAddress> agent_list(discovery_ports.size());
+        for(size_t i = 0; i < agent_list.size(); ++i)
+        {
+            strcpy(agent_list[i].ip, "127.0.0.1");
+            agent_list[i].port = uint16_t(discovery_ports[i] + i);
+        }
+
+        ASSERT_FALSE(uxr_discovery_agents_unicast(1, 1000, on_agent_found, this, &chosen_, agent_list.data(), agent_list.size()));
+        ASSERT_TRUE(agent_ports_.empty());
+    }
+
+    void multicast()
+    {
+        ASSERT_FALSE(uxr_discovery_agents_multicast(1, 1000, on_agent_found, this, &chosen_));
+        ASSERT_TRUE(agent_ports_.empty());
+    }
+
+private:
+    static bool on_agent_found(const uxrAgentAddress* address, int64_t timestamp, void* args)
+    {
+        static_cast<Discovery*>(args)->on_agent_found_member(address, timestamp);
+        return false;
+    }
+
+    void on_agent_found_member(const uxrAgentAddress* address, int64_t timestamp)
```

Comment:
Why is `timestamp` an `int64_t`?

Reply:
If the question is about why not an `unint64_t`. it is because the extra bit is lost. Sooner or later, this value can be cast to `int64_t` to make differences between time, and the last bit information will be lost.

Reply:
Ok!